<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate Sequencing Reads for Single Cells in Batches — simulate_sc_dynamic_reads_for_batches • cancerSimCraft</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Sequencing Reads for Single Cells in Batches — simulate_sc_dynamic_reads_for_batches"><meta property="og:description" content="Simulates sequencing reads for multiple single cells in batches by synthesizing
cell-specific genomes with mutations and generating sequencing reads using the
ART simulator. This function processes cells in batches to manage memory usage."><meta property="og:image" content="https://haijingjin.github.io/cancerSimCraft/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancerSimCraft</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basic_clonal_simulation_mini_genome.html">Basic Clonal Simulation with Mini Genome</a>
    </li>
    <li>
      <a href="../articles/advanced_1_chisel_data_preprocessing.html">Advanced 1: Chisel Data Preprocessing</a>
    </li>
    <li>
      <a href="../articles/advanced_2_pattern_extraction.html">Advanced 2: Pattern Extraction</a>
    </li>
    <li>
      <a href="../articles/advanced_3_genomic_region_mapping.html">Advanced 3: Genomic Region Mapping</a>
    </li>
    <li>
      <a href="../articles/advanced_4_simulation_setting.html">Advanced 4: Simulation Setting</a>
    </li>
    <li>
      <a href="../articles/advanced_5_backbone_genome_with_snp.html">Advanced 5: Build Backbone Genome with SNP</a>
    </li>
    <li>
      <a href="../articles/advanced_6_clonal_level_cancer_genome_simulation.html">Advanced 6: Clonal-level Cancer Genome Simulation</a>
    </li>
    <li>
      <a href="../articles/advanced_7_clonal_level_read_simulation.html">Advanced 7: Clonal-level Read Simulation</a>
    </li>
    <li>
      <a href="../articles/advanced_8_dynamics_simulation.html">Advanced 8: Single-cell Dynamics Simulation Setting</a>
    </li>
    <li>
      <a href="../articles/advanced_9_dynamics_genome_and_read_simulation.html">Advanced 9: Single-cell Dynamics Genome and Read Simulation</a>
    </li>
  </ul></li>
<li>
  <a href="../CITATION.html">Cite</a>
</li>
<li>
  <a href="../LICENSE.html">License</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Sequencing Reads for Single Cells in Batches</h1>
    <small class="dont-index">Source: <a href="https://github.com/haijingjin/cancerSimCraft/blob/HEAD/R/read_sim.R" class="external-link"><code>R/read_sim.R</code></a></small>
    <div class="hidden name"><code>simulate_sc_dynamic_reads_for_batches.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Simulates sequencing reads for multiple single cells in batches by synthesizing
cell-specific genomes with mutations and generating sequencing reads using the
ART simulator. This function processes cells in batches to manage memory usage.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simulate_sc_dynamic_reads_for_batches</span><span class="op">(</span></span>
<span>  <span class="va">sampled_cell_idx</span>,</span>
<span>  <span class="va">dynamics_ob</span>,</span>
<span>  <span class="va">sc_founder_genomes</span>,</span>
<span>  <span class="va">all_sampled_sc_mutations</span>,</span>
<span>  <span class="va">sc_mut_table_with_nt</span>,</span>
<span>  <span class="va">sim_updates</span>,</span>
<span>  <span class="va">fa_dir</span>,</span>
<span>  <span class="va">output_dir</span>,</span>
<span>  <span class="va">art_path</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  depth <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  readLen <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  otherArgs <span class="op">=</span> <span class="st">""</span>,</span>
<span>  keep_genome_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-sampled-cell-idx">sampled_cell_idx<a class="anchor" aria-label="anchor" href="#arg-sampled-cell-idx"></a></dt>
<dd><p>Vector of cell indices for which to simulate reads</p></dd>


<dt id="arg-dynamics-ob">dynamics_ob<a class="anchor" aria-label="anchor" href="#arg-dynamics-ob"></a></dt>
<dd><p>List object returned by simulate_sc_dynamics containing
cell lineage information</p></dd>


<dt id="arg-sc-founder-genomes">sc_founder_genomes<a class="anchor" aria-label="anchor" href="#arg-sc-founder-genomes"></a></dt>
<dd><p>List containing genome information for clone founders
with elements: all_node_genomes, child_node_founders_df</p></dd>


<dt id="arg-all-sampled-sc-mutations">all_sampled_sc_mutations<a class="anchor" aria-label="anchor" href="#arg-all-sampled-sc-mutations"></a></dt>
<dd><p>List containing mutation information for sampled cells</p></dd>


<dt id="arg-sc-mut-table-with-nt">sc_mut_table_with_nt<a class="anchor" aria-label="anchor" href="#arg-sc-mut-table-with-nt"></a></dt>
<dd><p>Data frame containing mutation details with nucleotide changes</p></dd>


<dt id="arg-sim-updates">sim_updates<a class="anchor" aria-label="anchor" href="#arg-sim-updates"></a></dt>
<dd><p>List containing updated simulation information including all_node_segments</p></dd>


<dt id="arg-fa-dir">fa_dir<a class="anchor" aria-label="anchor" href="#arg-fa-dir"></a></dt>
<dd><p>Character string specifying the directory to store FASTA files</p></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>Character string specifying the directory for output read files</p></dd>


<dt id="arg-art-path">art_path<a class="anchor" aria-label="anchor" href="#arg-art-path"></a></dt>
<dd><p>Character string specifying the path to the ART sequencing simulator executable</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Integer specifying the number of CPU cores to use for parallel processing (default: 4)</p></dd>


<dt id="arg-depth">depth<a class="anchor" aria-label="anchor" href="#arg-depth"></a></dt>
<dd><p>Numeric value specifying the desired sequencing depth/coverage (default: 30)</p></dd>


<dt id="arg-readlen">readLen<a class="anchor" aria-label="anchor" href="#arg-readlen"></a></dt>
<dd><p>Integer specifying the length of simulated reads in base pairs (default: 150)</p></dd>


<dt id="arg-otherargs">otherArgs<a class="anchor" aria-label="anchor" href="#arg-otherargs"></a></dt>
<dd><p>Character string with additional arguments to pass to ART (default: "")</p></dd>


<dt id="arg-keep-genome-files">keep_genome_files<a class="anchor" aria-label="anchor" href="#arg-keep-genome-files"></a></dt>
<dd><p>Logical indicating whether to retain the temporary FASTA
genome files after simulation (default: FALSE)</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Integer specifying the number of cells to process in each batch (default: 10)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list where each element contains mutation check results for a cell, with the
cell index as the name of each element</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function performs the following steps for each batch of cells:</p><ol><li><p>For each cell in the batch:</p><ul><li><p>Extracts cell lineage information from dynamics_ob</p></li>
<li><p>Synthesizes a cell-specific genome with mutations using synth_sc_genome()</p></li>
<li><p>Performs checks on the generated genome using check_genome_mutations()</p></li>
<li><p>Validates the sanity check results with validate_mutation_check()</p></li>
<li><p>Combines maternal and paternal haplotypes into a single FASTA file</p></li>
<li><p>Writes the merged genome to disk</p></li>
</ul></li>
<li><p>Simulates sequencing reads for all cells in the batch in parallel using simulate_read_dynamic_sc()</p></li>
<li><p>Optionally removes temporary genome files to save disk space</p></li>
</ol><p>Processing cells in batches helps manage memory usage when dealing with large numbers of cells.
The function relies on external functions like synth_sc_genome(), check_genome_mutations(),
validate_mutation_check(), and simulate_read_dynamic_sc().</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="synth_sc_genome.html">synth_sc_genome</a></code>, <code><a href="check_genome_mutations.html">check_genome_mutations</a></code>,
<code><a href="validate_mutation_check.html">validate_mutation_check</a></code>, <code><a href="simulate_read_dynamic_sc.html">simulate_read_dynamic_sc</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Assuming you have already run a simulation and have necessary objects</span></span></span>
<span class="r-in"><span><span class="va">mutation_checks</span> <span class="op">&lt;-</span> <span class="fu">simulate_sc_dynamic_reads_for_batches</span><span class="op">(</span></span></span>
<span class="r-in"><span>  sampled_cell_idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">51</span>, <span class="fl">52</span>, <span class="fl">53</span>, <span class="fl">54</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  dynamics_ob <span class="op">=</span> <span class="va">sc_dynamics_result</span>,</span></span>
<span class="r-in"><span>  sc_founder_genomes <span class="op">=</span> <span class="va">founder_genomes</span>,</span></span>
<span class="r-in"><span>  all_sampled_sc_mutations <span class="op">=</span> <span class="va">sampled_mutations</span>,</span></span>
<span class="r-in"><span>  sc_mut_table_with_nt <span class="op">=</span> <span class="va">mutations_with_nt</span>,</span></span>
<span class="r-in"><span>  sim_updates <span class="op">=</span> <span class="va">simulation_updates</span>,</span></span>
<span class="r-in"><span>  fa_dir <span class="op">=</span> <span class="st">"data/genomes/"</span>,</span></span>
<span class="r-in"><span>  output_dir <span class="op">=</span> <span class="st">"results/reads/"</span>,</span></span>
<span class="r-in"><span>  art_path <span class="op">=</span> <span class="st">"/usr/local/bin/art_illumina"</span>,</span></span>
<span class="r-in"><span>  n_cores <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>  depth <span class="op">=</span> <span class="fl">30</span>,</span></span>
<span class="r-in"><span>  readLen <span class="op">=</span> <span class="fl">150</span>,</span></span>
<span class="r-in"><span>  keep_genome_files <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  batch_size <span class="op">=</span> <span class="fl">2</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Check results for a specific cell</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">mutation_checks</span><span class="op">[[</span><span class="st">"50"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Haijing Jin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

