<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate Clonal-level Single-Cell Sequencing Reads — simulate_read_sc • cancerSimCraft</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Clonal-level Single-Cell Sequencing Reads — simulate_read_sc"><meta property="og:description" content="This function uses the ART read simulator (Huang et al., Bioinformatics 2012) to generate synthetic sequencing
reads from a reference FASTA file for multiple single cells in parallel. It supports
both paired-end and single-end reads and allows for parallel execution across multiple cores.
The function generates FASTQ files for each cell and logs the ART output."><meta property="og:image" content="https://haijingjin.github.io/cancerSimCraft/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancerSimCraft</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basic_clonal_simulation_mini_genome.html">Basic Clonal Simulation with Mini Genome</a>
    </li>
    <li>
      <a href="../articles/advanced_1_chisel_data_preprocessing.html">Advanced 1: Chisel Data Preprocessing</a>
    </li>
    <li>
      <a href="../articles/advanced_2_pattern_extraction.html">Advanced 2: Pattern Extraction</a>
    </li>
    <li>
      <a href="../articles/advanced_3_genomic_region_mapping.html">Advanced 3: Genomic Region Mapping</a>
    </li>
    <li>
      <a href="../articles/advanced_4_simulation_setting.html">Advanced 4: Simulation Setting</a>
    </li>
    <li>
      <a href="../articles/advanced_5_backbone_genome_with_snp.html">Advanced 5: Build Backbone Genome with SNP</a>
    </li>
    <li>
      <a href="../articles/advanced_6_clonal_level_cancer_genome_simulation.html">Advanced 6: Clonal-level Cancer Genome Simulation</a>
    </li>
    <li>
      <a href="../articles/advanced_7_clonal_level_read_simulation.html">Advanced 7: Clonal-level Read Simulation</a>
    </li>
    <li>
      <a href="../articles/advanced_8_dynamics_simulation.html">Advanced 8: Single-cell Dynamics Simulation Setting</a>
    </li>
    <li>
      <a href="../articles/advanced_9_dynamics_genome_and_read_simulation.html">Advanced 9: Single-cell Dynamics Genome and Read Simulation</a>
    </li>
  </ul></li>
<li>
  <a href="../CITATION.html">Cite</a>
</li>
<li>
  <a href="../LICENSE.html">License</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Clonal-level Single-Cell Sequencing Reads</h1>
    <small class="dont-index">Source: <a href="https://github.com/haijingjin/cancerSimCraft/blob/HEAD/R/read_sim.R" class="external-link"><code>R/read_sim.R</code></a></small>
    <div class="hidden name"><code>simulate_read_sc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function uses the ART read simulator (Huang et al., Bioinformatics 2012) to generate synthetic sequencing
reads from a reference FASTA file for multiple single cells in parallel. It supports
both paired-end and single-end reads and allows for parallel execution across multiple cores.
The function generates FASTQ files for each cell and logs the ART output.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simulate_read_sc</span><span class="op">(</span></span>
<span>  <span class="va">fasta_input</span>,</span>
<span>  <span class="va">output_prefixes</span>,</span>
<span>  readLen <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  depth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  artPath <span class="op">=</span> <span class="st">"art_illumina"</span>,</span>
<span>  seqSys <span class="op">=</span> <span class="st">"HS25"</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  numCores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  otherArgs <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-fasta-input">fasta_input<a class="anchor" aria-label="anchor" href="#arg-fasta-input"></a></dt>
<dd><p>Character string. Path to the input FASTA file.</p></dd>


<dt id="arg-output-prefixes">output_prefixes<a class="anchor" aria-label="anchor" href="#arg-output-prefixes"></a></dt>
<dd><p>Character vector. Prefixes for output files, one per simulated cell.</p></dd>


<dt id="arg-readlen">readLen<a class="anchor" aria-label="anchor" href="#arg-readlen"></a></dt>
<dd><p>Numeric. Length of the simulated reads in base pairs. Default is 100.</p></dd>


<dt id="arg-depth">depth<a class="anchor" aria-label="anchor" href="#arg-depth"></a></dt>
<dd><p>Numeric. Sequencing coverage depth. Default is 1.</p></dd>


<dt id="arg-artpath">artPath<a class="anchor" aria-label="anchor" href="#arg-artpath"></a></dt>
<dd><p>Character string. Path to the ART Illumina executable. Default is "art_illumina".</p></dd>


<dt id="arg-seqsys">seqSys<a class="anchor" aria-label="anchor" href="#arg-seqsys"></a></dt>
<dd><p>Character string. Sequencing system to simulate. Default is "HS25" (HiSeq 2500).</p></dd>


<dt id="arg-paired">paired<a class="anchor" aria-label="anchor" href="#arg-paired"></a></dt>
<dd><p>Logical. Whether to generate paired-end reads (TRUE) or single-end reads (FALSE). Default is TRUE.</p></dd>


<dt id="arg-numcores">numCores<a class="anchor" aria-label="anchor" href="#arg-numcores"></a></dt>
<dd><p>Numeric. Number of cores to use for parallel processing. Default is 2.</p></dd>


<dt id="arg-otherargs">otherArgs<a class="anchor" aria-label="anchor" href="#arg-otherargs"></a></dt>
<dd><p>Character string. Additional arguments to pass to ART Illumina. Default is "".</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>No return value, called for side effects of generating simulated read files and
printing completion messages.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function creates synthetic sequencing reads by calling ART Illumina for each output prefix
in parallel using mclapply. For paired-end reads, it sets a mean fragment length of 200bp with
a standard deviation of 10bp. Simulation output and errors are redirected to log files.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Simulate reads for 3 cells with default parameters</span></span></span>
<span class="r-in"><span><span class="fu">simulate_read_sc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  fasta_input <span class="op">=</span> <span class="st">"reference.fa"</span>,</span></span>
<span class="r-in"><span>  output_prefixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1_"</span>, <span class="st">"cell2_"</span>, <span class="st">"cell3_"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate single-end reads with custom parameters</span></span></span>
<span class="r-in"><span><span class="fu">simulate_read_sc</span><span class="op">(</span></span></span>
<span class="r-in"><span>  fasta_input <span class="op">=</span> <span class="st">"reference.fa"</span>,</span></span>
<span class="r-in"><span>  output_prefixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1_"</span>, <span class="st">"cell2_"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  readLen <span class="op">=</span> <span class="fl">150</span>,</span></span>
<span class="r-in"><span>  depth <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>  paired <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  <span class="va">...</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Haijing Jin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

